### ISSUE TITLE
Review of the current v2.0 branch

### ISSUE THREAD

**Issue by giumas** [OpenNavigationSurface/BAG#51]
The following comments/suggestions are based on the code available in https://github.com/OpenNavigationSurface/BAG/commit/b7f024a5ed246339d30318c9a5c5daae8db3f72f 

General
---------

* The `CHANGES.txt` file is outdated. The large changes in release 2.0 should be briefly described there.

* The `docs\build?instructions.html` file is outdated. The new instructions should be much shorter.

* The `examples\readme.txt` file is outdated. There are existing example scripts that are missing (`bag_vr_create.cpp`), and there are listed scripts that are not actually present (e.g., `bag_signfile`).

* How to build the doxygen documentation?

* The encryption with _beecrypt_ is absent. Is it because of the decision to have a stand-alone tool to do this task? If so, why do we still have encryption-related errors? e.g., `BAG_CRYPTO_NO_SIGNATURE_FOUND` in `bag_errors.h`?

* What is the current use of the `extlibs` folder?

CMake
--------

* Why enforcing the maximum version of CMake to 3.16? CMake 3.17 has already been released.  I understand the need of a minimum version. Thus, I would suggest to remove the `...3.16`:
   * `cmake_minimum_required(VERSION 3.0...3.16)` (see https://cmake.org/cmake/help/latest/command/cmake_minimum_required.html)

Catch2 tests
-------------

* The catch2 tests fail to run within Visual Studio because they don't find the BAG dll. Manually copying the BAG dll is annoying and error-prone:

![u6tcdmKNJG](https://user-images.githubusercontent.com/2849257/81498023-bd921680-92c2-11ea-943c-bed6c47fe5c4.png)

* More generally, how running the tests from Visual Studio? It does not seem that Visual Studio recognize them as tests by default (as it happens with other testing frameworks). If it does, some guidance on how to do it would help.

C++ Library
-------------

* The concept of `interleaved layers` should be clearly presented/described. The selected name does not really help to understand what they are. I would suggest to add `legacy` (or something similar) to such a name. It would also be useful to add a reference to the ONS meeting when the unpacking of the NODE group and the ELEVATION group was discussed and approved.

Python binding
-----------------

* The build of the Python binding currently requires to run Visual Studio as administrator. Otherwise there is a CMake error:  `CMake Error: failed to create symbolic link: operation not permitted`. I suggest to copy the required files rather than creating a symbolic link. 

* The Python binding should be built out of source: _'One nice and highly recommended feature of CMake is the ability to do out of source builds. In this way you can make all your .o files, various temporary depend files, and even the binary executables without cluttering up your source tree.'_ (see https://www.cs.swarthmore.edu/~adanner/tips/cmake.php).

* The CMake install command ignores the Python binding. What is the current way to install the Python binding? I was not able to find any `setup.py` (https://packaging.python.org/tutorials/packaging-projects/). It would also be useful to put the module under a namespace (e.g., `ons`) to avoid to pollute the `site-packages` folder.

* The Python tests do not use a testing framework. The current hard-coded solution of using `assert()` is difficult to maintain as well as unusual for Python libraries. Python has a built-in testing framework named `unittest` (https://docs.python.org/3.6/library/unittest.html) very simple to use.

* There is NOT documentation on how to use the Python binding. The Python tests can be somehow useful, but they cannot fully substitute the need of proper documentation. The most popular way to document Python libraries is by using Sphynx (https://www.sphinx-doc.org/en/master/).

Continuous integration
-------------------------

* Travis and AppVeyor tests Python 3.6 and 3.7. The current version of Python is 3.8, thus it should be also added.

* Travis script has developer-specific calls (`LD_LIBRARY_PATH=/home/travis/build/jice73/BAG/build/api PYTHONPATH=/home/travis/build/jice73/BAG/build/api/swig_i python python/test_all.py`) that should be removed. Otherwise, how is this going to work when executed by the upstream GtiHub repository?

CC: @GlenRice-NOAA @jice73 @brian-r-calder

**Comment by GlenRice-NOAA** [comment]
@johnsonst, do you have a requirement for the `extlibs` folder?  I think this can be removed from the library and added and referenced if needed on a case by case basis.  It doesn't seem like this should be part of the library anymore.

**Comment by GlenRice-NOAA** [comment]
@johnsonst, do you have a minimum version of CMake you require for your environment?

**Comment by johnsonst** [comment]
@GlenRice-NOAA 

We currently have access to cmake 2.8.12.2 so we need support at least back to that version.

I don't think extlibs needs to be part of the primary download package.  I would like to keep it around as an ancillary download, if possible.

**Comment by giumas** [comment]
@johnsonst CMake 2.8.12.2 does not support Visual Studio 2015 (it stops at Visual Studio 2012) on Windows. The code has now several C++11 features in use. 

However, the library may actually compile outside of Visual Studio realm. Have you tried to compile this branch on your workstation after having modified the minimum required CMake version?

**Comment by johnsonst** [comment]
@giumas We don't control our version of CMake.  It is whatever is allowed by the DoD.  For compilers and development tools on Linux, this is whatever RHEL supplied via their package servers.  If we deviate from this, it causes IA related issues.

That said, we usually do not use CMake on the Linux side.  We just use the standard Makefile.  This can be done without modifications to the downloaded code.  Is Visual Studio 15 required to compile the code via CMake?  If not, I would prefer to not define a minimum version of CMake based on a tool (Visual Studio) that uses CMake rather that the actual CMake requirement.

**Comment by johnsonst** [comment]
@giumas We don't control our version of CMake.  It is whatever is allowed by the DoD.  For compilers and development tools on Linux, this is whatever RHEL supplied via their package servers.  If we deviate from this, it causes IA related issues.

That said, we usually do not use CMake on the Linux side.  We just use the standard Makefile.  This can be done without modifications to the downloaded code.  Is Visual Studio 15 required to compile the code via CMake?  If not, I would prefer to not define a minimum version of CMake based on a tool (Visual Studio) that uses CMake rather that the actual CMake requirement.

**Comment by giumas** [comment]
@johnsonst Understood! Just a curiosity, do you use your own makefile to build the BAG library? I cannot find it in the zipped 1.6.3 (http://opennavsurf.org/dl/send_package.php)

**Comment by johnsonst** [comment]
@giumas Yes, we have our own Makefile.  We have to substitute some of dependencies like zlib because of IA issues, so we end up with a custom file either way,  

I think a Makefile was part of the distribution in the early v1.5.X or v1.4.X versions but officially it moved to CMake a few years back.
